src/plugins/discord/mirror.js

```diff
@@ -0,0 +1,78 @@
+// @flow
+
+import {type ResultPage} from "./fetcher";
+import {type SqliteMirror} from "./sqliteMirror";
+import {type Snowflake} from "./models";
+import * as Model from "./models";
+import * as nullUtil from "../../util/null";
+
+/**
+ * Mirrors data from the Discord API into a local sqlite db.
+ */
+
+export async function fetchDiscord(
+  sqliteMirror: SqliteMirror,
+  streams: StreamingFetcher
+) {
+  for (const member of await streams.members()) {
+    sqliteMirror.addMember(member);
+  }
+
+  for (const channel of await streams.channels()) {
+    sqliteMirror.addChannel(channel);
+
+    for (const message of await streams.messages(channel.id)) {
+      sqliteMirror.addMessage(message);
+
+      for (const emoji of message.reactionEmoji) {
+        for (const reaction of await streams.reactions(
+          channel.id,
+          message.id,
+          emoji
+        )) {
+          sqliteMirror.addReaction(reaction);
+        }
+      }
+    }
+  }
+}
+
+// Note: most of this is about wrapping pagination.
+export interface StreamingFetcher {
+  members(after: ?Snowflake): Promise<$ReadOnlyArray<Model.GuildMember>>;
+  channels(): Promise<$ReadOnlyArray<Model.Channel>>;
+  messages(
+    channel: Snowflake,
+    after: ?Snowflake
+  ): Promise<$ReadOnlyArray<Model.Message>>;
+  reactions(
+    channel: Snowflake,
+    message: Snowflake,
+    emoji: Model.Emoji,
+    after: ?Snowflake
+  ): Promise<$ReadOnlyArray<Model.Reaction>>;
+}
+
+export async function getPages<T>(
+  fetchPage: (endCursor: Snowflake) => Promise<ResultPage<T>>,
+  endCursor: Snowflake
+): Promise<$ReadOnlyArray<T>> {
+  let after = endCursor;
+  let hasNextPage = true;
+  let allResults = [];
```

Comment:
This implementation accumulates results in this variable. The upside is it makes things very simple. The tradeoff may be that channels have tens of thousands of messages. It's not a lot of added complexity to pass in a callback that processes these results individually, instead of accumulating them all and returning them. However, opting for the simplest version in this pass.

Reply:
Cool. Yeah an async generator would conceptually fit well here, but fine with this implementation for now.